File: apps/web/src/lib/api/masterclass.ts
/**
 * Masterclass API Client
 * 
 * Client for fetching masterclass events, cities, and availability data
 */

import { apiClient, extractApiData } from '@/lib/api';
import type { ApiResponse } from '@modele/types';

export interface MasterclassEvent {
  id: number;
  title_en: string;
  title_fr: string;
  description_en?: string;
  description_fr?: string;
  duration_days: number;
  language: string;
  created_at: string;
  updated_at: string;
  title?: string;  // Alias for compatibility (will map to title_fr or title_en)
  description?: string;  // Alias for compatibility (will map to description_fr or description_en)
}

export interface City {
  id: number;
  name_en: string;
  name_fr: string;
  name?: string;  // Alias for compatibility (will map to name_fr or name_en)
  country: string;
  province?: string;
  timezone?: string;
  image_url?: string;
}

export interface Venue {
  id: number;
  name: string;
  address: string;
  city_id: number;
  capacity?: number;
  amenities?: string[];
}

export interface CityEvent {
  id: number;
  event_id: number;
  city_id: number;
  venue_id: number;
  start_date: string;  // Primary date field
  end_date: string;
  event_date?: string;  // Optional alias for start_date (for compatibility)
  start_time: string;
  end_time: string;
  total_capacity?: number;
  available_spots?: number;
  max_attendees?: number;  // Alias for total_capacity
  current_attendees?: number;  // Calculated: total_capacity - available_spots
  is_active?: boolean;  // status == PUBLISHED
  status?: string;  // draft, published, sold_out, cancelled
  price?: number;  // Alias for regular_price
  currency?: string;
  regular_price?: number;
  early_bird_price?: number;
  early_bird_deadline?: string;
  group_discount_percentage?: number;
  group_minimum?: number;
  event?: MasterclassEvent;
  city?: City;
  venue?: Venue;
}

export interface Availability {
  total_spots: number;
  available_spots: number;
  booked_spots: number;
  is_available: boolean;
  percentage_available: number;
}

export interface CityWithEvents {
  id: number;
  name_en: string;
  name_fr: string;
  name?: string;  // Alias
  country: string;
  province?: string;
  timezone?: string;
  events?: CityEvent[];
  city_events?: CityEvent[];  // Alias from backend
}

/**
 * Masterclass API client
 */
export const masterclassAPI = {
  /**
   * Get list of all masterclass events
   */
  listEvents: async (): Promise<MasterclassEvent[]> => {
    const response = await apiClient.get<MasterclassEvent[]>('/v1/masterclass/events');
    const data = extractApiData<MasterclassEvent[]>(response as unknown as ApiResponse<MasterclassEvent[]>);
    return Array.isArray(data) ? data : [];
  },

  /**
   * Get a specific event by ID
   */
  getEvent: async (eventId: number): Promise<MasterclassEvent> => {
    const response = await apiClient.get<MasterclassEvent>(`/v1/masterclass/events/${eventId}`);
    const data = extractApiData<MasterclassEvent>(response as unknown as ApiResponse<MasterclassEvent>);
    if (!data) {
      throw new Error(`Event not found: ${eventId}`);
    }
    return data;
  },

  /**
   * Get list of cities with their events
   */
  listCitiesWithEvents: async (): Promise<CityWithEvents[]> => {
    const response = await apiClient.get<{cities: CityWithEvents[], total: number}>('/v1/masterclass/cities');
    const result = extractApiData<{cities: CityWithEvents[], total: number}>(response as unknown as ApiResponse<{cities: CityWithEvents[], total: number}>);
    if (!result || !result.cities) {
      return [];
    }
    // Map city_events to events for frontend compatibility
    return result.cities.map(city => ({
      ...city,
      name: city.name_fr || city.name_en,
      events: city.city_events || city.events || [],
    }));
  },

  /**
   * Get events for a specific city
   */
  listCityEvents: async (cityId: number): Promise<CityEvent[]> => {
    const response = await apiClient.get<{city_events: CityEvent[], total: number}>(`/v1/masterclass/cities/${cityId}/events`);
    const result = extractApiData<{city_events: CityEvent[], total: number}>(response as unknown as ApiResponse<{city_events: CityEvent[], total: number}>);
    if (!result || !result.city_events) {
      return [];
    }
    // Map fields for frontend compatibility
    return result.city_events.map(event => ({
      ...event,
      event_date: event.start_date,
      max_attendees: event.total_capacity,
      current_attendees: (event.total_capacity || 0) - (event.available_spots || 0),
      is_active: event.status === 'published',
      price: event.regular_price,
    }));
  },

  /**
   * Get a specific city event by ID
   */
  getCityEvent: async (cityEventId: number): Promise<CityEvent> => {
    const response = await apiClient.get<CityEvent>(`/v1/masterclass/city-events/${cityEventId}`);
    const data = extractApiData<CityEvent>(response as unknown as ApiResponse<CityEvent>);
    if (!data) {
      throw new Error(`City event not found: ${cityEventId}`);
    }
    // Map fields for frontend compatibility
    return {
      ...data,
      event_date: data.start_date,
      max_attendees: data.total_capacity,
      current_attendees: (data.total_capacity || 0) - (data.available_spots || 0),
      is_active: data.status === 'published',
      price: data.regular_price,
      currency: data.currency || 'EUR',
    };
  },

  /**
   * Get availability for a city event
   */
  getAvailability: async (cityEventId: number): Promise<Availability> => {
    const response = await apiClient.get<Availability>(`/v1/masterclass/city-events/${cityEventId}/availability`);
    const data = extractApiData<Availability>(response as unknown as ApiResponse<Availability>);
    if (!data) {
      throw new Error(`Availability not found for city event: ${cityEventId}`);
    }
    return data;
  },

  // ============================================================================
  // Admin endpoints (require admin or superadmin)
  // ============================================================================

  /**
   * Create a new masterclass event (admin only)
   */
  createEvent: async (data: {
    title_en: string;
    title_fr: string;
    description_en?: string;
    description_fr?: string;
    duration_days?: number;
    language?: string;
  }): Promise<MasterclassEvent> => {
    const response = await apiClient.post<MasterclassEvent>('/v1/masterclass/events', data);
    return extractApiData<MasterclassEvent>(response as unknown as ApiResponse<MasterclassEvent>);
  },

  /**
   * Update a masterclass event (admin only)
   */
  updateEvent: async (eventId: number, data: {
    title_en?: string;
    title_fr?: string;
    description_en?: string;
    description_fr?: string;
    duration_days?: number;
    language?: string;
  }): Promise<MasterclassEvent> => {
    const response = await apiClient.put<MasterclassEvent>(`/v1/masterclass/events/${eventId}`, data);
    return extractApiData<MasterclassEvent>(response as unknown as ApiResponse<MasterclassEvent>);
  },

  /**
   * Delete a masterclass event (admin only)
   */
  deleteEvent: async (eventId: number): Promise<void> => {
    await apiClient.delete(`/v1/masterclass/events/${eventId}`);
  },

  /**
   * List all cities (admin only, without filtering)
   */
  listAllCities: async (): Promise<City[]> => {
    const response = await apiClient.get<City[]>('/v1/masterclass/cities/all');
    const data = extractApiData<City[]>(response as unknown as ApiResponse<City[]>);
    return Array.isArray(data) ? data : [];
  },

  /**
   * Create a new city (admin only)
   */
  createCity: async (data: {
    name_en: string;
    name_fr: string;
    province?: string;
    country?: string;
    timezone?: string;
    image_url?: string;
  }): Promise<City> => {
    const response = await apiClient.post<City>('/v1/masterclass/cities', data);
    return extractApiData<City>(response as unknown as ApiResponse<City>);
  },

  /**
   * Update a city (admin only)
   */
  updateCity: async (cityId: number, data: {
    name_en?: string;
    name_fr?: string;
    province?: string;
    country?: string;
    timezone?: string;
    image_url?: string;
  }): Promise<City> => {
    const response = await apiClient.put<City>(`/v1/masterclass/cities/${cityId}`, data);
    return extractApiData<City>(response as unknown as ApiResponse<City>);
  },

  /**
   * Delete a city (admin only)
   */
  deleteCity: async (cityId: number): Promise<void> => {
    await apiClient.delete(`/v1/masterclass/cities/${cityId}`);
  },

  /**
   * List all venues (admin only)
   */
  listVenues: async (cityId?: number): Promise<Venue[]> => {
    const response = await apiClient.get<Venue[]>('/v1/masterclass/venues', {
      params: cityId ? { city_id: cityId } : undefined,
    });
    const data = extractApiData<Venue[]>(response as unknown as ApiResponse<Venue[]>);
    return Array.isArray(data) ? data : [];
  },

  /**
   * Get a venue by ID (admin only)
   */
  getVenue: async (venueId: number): Promise<Venue> => {
    const response = await apiClient.get<Venue>(`/v1/masterclass/venues/${venueId}`);
    const data = extractApiData<Venue>(response as unknown as ApiResponse<Venue>);
    if (!data) {
      throw new Error(`Venue not found: ${venueId}`);
    }
    return data;
  },

  /**
   * Create a new venue (admin only)
   */
  createVenue: async (data: {
    city_id: number;
    name: string;
    address?: string;
    postal_code?: string;
    capacity: number;
    amenities?: any;
  }): Promise<Venue> => {
    const response = await apiClient.post<Venue>('/v1/masterclass/venues', data);
    return extractApiData<Venue>(response as unknown as ApiResponse<Venue>);
  },

  /**
   * Update a venue (admin only)
   */
  updateVenue: async (venueId: number, data: {
    city_id?: number;
    name?: string;
    address?: string;
    postal_code?: string;
    capacity?: number;
    amenities?: any;
  }): Promise<Venue> => {
    const response = await apiClient.put<Venue>(`/v1/masterclass/venues/${venueId}`, data);
    return extractApiData<Venue>(response as unknown as ApiResponse<Venue>);
  },

  /**
   * Delete a venue (admin only)
   */
  deleteVenue: async (venueId: number): Promise<void> => {
    await apiClient.delete(`/v1/masterclass/venues/${venueId}`);
  },

  /**
   * List all city events (admin only, without filtering)
   */
  listAllCityEvents: async (skip = 0, limit = 100, status?: string): Promise<{city_events: CityEvent[], total: number}> => {
    const response = await apiClient.get<{city_events: CityEvent[], total: number}>('/v1/masterclass/city-events/all', {
      params: { skip, limit, status_filter: status },
    });
    const result = extractApiData<{city_events: CityEvent[], total: number}>(response as unknown as ApiResponse<{city_events: CityEvent[], total: number}>);
    if (!result) {
      return { city_events: [], total: 0 };
    }
    // Map fields for frontend compatibility
    return {
      city_events: result.city_events.map(event => ({
        ...event,
        event_date: event.start_date,
        max_attendees: event.total_capacity,
        current_attendees: (event.total_capacity || 0) - (event.available_spots || 0),
        is_active: event.status === 'published',
        price: event.regular_price,
        currency: event.currency || 'EUR',
      })),
      total: result.total,
    };
  },

  /**
   * Create a new city event (admin only)
   */
  createCityEvent: async (data: {
    event_id: number;
    city_id: number;
    venue_id: number;
    start_date: string;
    end_date: string;
    start_time?: string;
    end_time?: string;
    total_capacity: number;
    available_spots: number;
    status?: string;
    early_bird_deadline?: string;
    early_bird_price?: number;
    regular_price: number;
    group_discount_percentage?: number;
    group_minimum?: number;
  }): Promise<CityEvent> => {
    const response = await apiClient.post<CityEvent>('/v1/masterclass/city-events', data);
    const event = extractApiData<CityEvent>(response as unknown as ApiResponse<CityEvent>);
    // Map fields for frontend compatibility
    return {
      ...event,
      event_date: event.start_date,
      max_attendees: event.total_capacity,
      current_attendees: (event.total_capacity || 0) - (event.available_spots || 0),
      is_active: event.status === 'published',
      price: event.regular_price,
      currency: event.currency || 'EUR',
    };
  },

  /**
   * Update a city event (admin only)
   */
  updateCityEvent: async (cityEventId: number, data: {
    event_id?: number;
    city_id?: number;
    venue_id?: number;
    start_date?: string;
    end_date?: string;
    start_time?: string;
    end_time?: string;
    total_capacity?: number;
    available_spots?: number;
    status?: string;
    early_bird_deadline?: string;
    early_bird_price?: number;
    regular_price?: number;
    group_discount_percentage?: number;
    group_minimum?: number;
  }): Promise<CityEvent> => {
    const response = await apiClient.put<CityEvent>(`/v1/masterclass/city-events/${cityEventId}`, data);
    const event = extractApiData<CityEvent>(response as unknown as ApiResponse<CityEvent>);
    // Map fields for frontend compatibility
    return {
      ...event,
      event_date: event.start_date,
      max_attendees: event.total_capacity,
      current_attendees: (event.total_capacity || 0) - (event.available_spots || 0),
      is_active: event.status === 'published',
      price: event.regular_price,
      currency: event.currency || 'EUR',
    };
  },

  /**
   * Delete a city event (admin only)
   */
  deleteCityEvent: async (cityEventId: number): Promise<void> => {
    await apiClient.delete(`/v1/masterclass/city-events/${cityEventId}`);
  },
};

File: backend/app/schemas/masterclass.py
"""
Masterclass Schemas
Pydantic schemas for masterclass events, cities, and venues
"""

from typing import Optional, List, Any
from datetime import datetime, date, time
from decimal import Decimal
from pydantic import BaseModel, Field, model_validator

from app.models.masterclass import EventStatus


# MasterclassEvent Schemas
class MasterclassEventBase(BaseModel):
    title_en: str = Field(..., max_length=200)
    title_fr: str = Field(..., max_length=200)
    description_en: Optional[str] = None
    description_fr: Optional[str] = None
    duration_days: int = Field(default=2, ge=1, le=30)
    language: str = Field(default="English", max_length=50)


class MasterclassEventCreate(MasterclassEventBase):
    pass


class MasterclassEventUpdate(BaseModel):
    title_en: Optional[str] = Field(None, max_length=200)
    title_fr: Optional[str] = Field(None, max_length=200)
    description_en: Optional[str] = None
    description_fr: Optional[str] = None
    duration_days: Optional[int] = Field(None, ge=1, le=30)
    language: Optional[str] = Field(None, max_length=50)


class MasterclassEventResponse(MasterclassEventBase):
    id: int
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


# City Schemas
class CityBase(BaseModel):
    name_en: str = Field(..., max_length=100)
    name_fr: str = Field(..., max_length=100)
    province: Optional[str] = Field(None, max_length=50)
    country: str = Field(default="Canada", max_length=50)
    timezone: str = Field(default="America/Toronto", max_length=50)
    image_url: Optional[str] = None


class CityCreate(CityBase):
    pass


class CityUpdate(BaseModel):
    name_en: Optional[str] = Field(None, max_length=100)
    name_fr: Optional[str] = Field(None, max_length=100)
    province: Optional[str] = Field(None, max_length=50)
    country: Optional[str] = Field(None, max_length=50)
    timezone: Optional[str] = Field(None, max_length=50)
    image_url: Optional[str] = None


class CityResponse(CityBase):
    id: int
    created_at: datetime

    class Config:
        from_attributes = True


# Venue Schemas
class VenueBase(BaseModel):
    city_id: int
    name: str = Field(..., max_length=200)
    address: Optional[str] = Field(None, max_length=300)
    postal_code: Optional[str] = Field(None, max_length=20)
    capacity: int = Field(..., ge=1)
    amenities: Optional[dict] = None  # JSON object


class VenueCreate(VenueBase):
    pass


class VenueUpdate(BaseModel):
    city_id: Optional[int] = None
    name: Optional[str] = Field(None, max_length=200)
    address: Optional[str] = Field(None, max_length=300)
    postal_code: Optional[str] = Field(None, max_length=20)
    capacity: Optional[int] = Field(None, ge=1)
    amenities: Optional[dict] = None


class VenueResponse(VenueBase):
    id: int
    created_at: datetime

    class Config:
        from_attributes = True


# CityEvent Schemas
class CityEventBase(BaseModel):
    event_id: int
    city_id: int
    venue_id: int
    start_date: date
    end_date: date
    start_time: time = Field(default=time(9, 0))
    end_time: time = Field(default=time(17, 0))
    total_capacity: int = Field(default=30, ge=1)
    available_spots: int = Field(..., ge=0)
    status: EventStatus = EventStatus.DRAFT
    early_bird_deadline: Optional[date] = None
    early_bird_price: Optional[Decimal] = Field(None, ge=0)
    regular_price: Decimal = Field(..., ge=0)
    group_discount_percentage: int = Field(default=10, ge=0, le=100)
    group_minimum: int = Field(default=3, ge=2)


class CityEventCreate(CityEventBase):
    pass


class CityEventUpdate(BaseModel):
    event_id: Optional[int] = None
    city_id: Optional[int] = None
    venue_id: Optional[int] = None
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    start_time: Optional[time] = None
    end_time: Optional[time] = None
    total_capacity: Optional[int] = Field(None, ge=1)
    available_spots: Optional[int] = Field(None, ge=0)
    status: Optional[EventStatus] = None
    early_bird_deadline: Optional[date] = None
    early_bird_price: Optional[Decimal] = Field(None, ge=0)
    regular_price: Optional[Decimal] = Field(None, ge=0)
    group_discount_percentage: Optional[int] = Field(None, ge=0, le=100)
    group_minimum: Optional[int] = Field(None, ge=2)


class CityEventResponse(CityEventBase):
    id: int
    created_at: datetime
    updated_at: datetime
    event: Optional[MasterclassEventResponse] = None
    city: Optional[CityResponse] = None
    venue: Optional[VenueResponse] = None
    # Aliases for frontend compatibility
    event_date: Optional[date] = None  # Alias for start_date
    max_attendees: Optional[int] = None  # Alias for total_capacity
    current_attendees: Optional[int] = None  # Calculated: total_capacity - available_spots
    is_active: Optional[bool] = None  # Alias for status == PUBLISHED
    price: Optional[Decimal] = None  # Alias for regular_price
    currency: Optional[str] = None  # Default to EUR

    class Config:
        from_attributes = True

    @model_validator(mode='before')
    @classmethod
    def add_computed_fields(cls, data: Any) -> Any:
        """Add computed fields for frontend compatibility"""
        if isinstance(data, dict):
            # Add computed fields
            if 'start_date' in data:
                data['event_date'] = data['start_date']
            if 'total_capacity' in data:
                data['max_attendees'] = data['total_capacity']
                # Calculate current_attendees
                total_cap = data.get('total_capacity', 0) or 0
                available = data.get('available_spots', 0) or 0
                data['current_attendees'] = max(0, total_cap - available)
            if 'status' in data:
                data['is_active'] = data['status'] == EventStatus.PUBLISHED
            if 'regular_price' in data:
                data['price'] = data['regular_price']
            data['currency'] = data.get('currency', 'EUR')
        elif hasattr(data, '__dict__'):
            # SQLAlchemy model - convert to dict
            obj_dict = {k: v for k, v in data.__dict__.items() if not k.startswith('_')}
            # Add computed fields
            if hasattr(data, 'start_date'):
                obj_dict['event_date'] = data.start_date
            if hasattr(data, 'total_capacity'):
                obj_dict['max_attendees'] = data.total_capacity
                total_cap = data.total_capacity or 0
                available = data.available_spots or 0
                obj_dict['current_attendees'] = max(0, total_cap - available)
            if hasattr(data, 'status'):
                obj_dict['is_active'] = data.status == EventStatus.PUBLISHED
            if hasattr(data, 'regular_price'):
                obj_dict['price'] = data.regular_price
            obj_dict['currency'] = 'EUR'
            return obj_dict
        return data


class CityWithEventsResponse(CityResponse):
    city_events: List[CityEventResponse] = []

    class Config:
        from_attributes = True


class AvailabilityResponse(BaseModel):
    """Availability response for a city event"""
    city_event_id: int
    total_capacity: int
    available_spots: int
    booked_spots: int
    percentage_available: float = Field(..., ge=0, le=100)
    status: str  # "available", "almost_full" (< 20%), "sold_out"
    is_almost_full: bool  # True if < 20% available
    is_sold_out: bool  # True if available_spots == 0


class EventListResponse(BaseModel):
    events: List[MasterclassEventResponse]
    total: int


class CityListResponse(BaseModel):
    cities: List[CityWithEventsResponse]
    total: int


class CityEventListResponse(BaseModel):
    city_events: List[CityEventResponse]
    total: int
